### DISCUSSION TITLE
Caddy brew service fails to start

### DISCUSSION THREAD

**Discussion by necrogami** [Homebrew/discussions#618] [Everyday usage]
So I've been working with the guys from caddy to debug an issue with the current caddy install.

For some reason when launching caddy with brew services it's trying to set the home directory inside /System/Volumes/Data which is read only.

When I run the exact same command in terminal that it runs from the plist it works fine. It's using the internal ssl built into caddy. Any idea how I can correct this?

**Comment by jonchang** [comment]
> For some reason when launching caddy with brew services it's trying to set the home directory inside /System/Volumes/Data which is read only. When I run the exact same command in terminal that it runs from the plist it works fine. It's using the internal ssl built into caddy. Any idea how I can correct this?

I believe this is by design when running things with `launchctl`, which is what `brew services` uses. Read `man launchd.plist` to see how to change a service's working directory.

**Reply by necrogami** [reply]
I understand how to change working directory, however this is setting the HOME directory to sys/vol/data

**Reply by jonchang** [reply]
Can you open an actual issue in homebrew/core and fill out all the details? Thanks.

**Reply by carlocab** [reply]
You can set the `HOME` directory manually using `EnvironmentVariables` (see `man launchd.plist`). Not sure why it's picking that `HOME` directory to start with though.

**Reply by necrogami** [reply]
Except when trying to set HOME using 
```
    <key>EnvironmentVariables</key>
    <dict>
        <key>HOME</key>
        <string>/usr/local/etc/caddy</string>
    </dict>
```

Caddy still returns:
```
{"level":"warn","ts":1611737159.153505,"msg":"unable to determine directory for user configuration; falling back to current directory","error":"$HOME is not defined"}
{"level":"info","ts":1611737159.173214,"msg":"using provided configuration","config_file":"/usr/local/etc/Caddyfile","config_adapter":""}
{"level":"info","ts":1611737159.175015,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]}
{"level":"info","ts":1611737159.1752791,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc000350690"}
{"level":"info","ts":1611737159.1836858,"logger":"tls.cache.maintenance","msg":"stopped background certificate maintenance","cache":"0xc000350690"}
run: loading initial config: loading new config: loading http app module: provision http: getting tls app: loading tls app module: provision tls: provisioning automation policy 0: loading TLS automation management module: position 0: loading module 'internal': provision tls.issuance.internal: loading pki app module: provision pki: provisioning CA 'local': generating root: saving root certificate: mkdir caddy: read-only file system
```

**Reply by SMillerDev** [reply]
Did you reload the plist file? If you change the homebrew version that will mean stopping, removing, reinstalling and restarting the formula.

**Reply by necrogami** [reply]
I finally got it to work, set the env variable for HOME to / and it finally loaded the server. This however presents a problem as when using caddy's internal ssl generation it generates a number of pki files those are part of what it is trying to create structure for when starting the server. Obviously using / for home isn't an ideal solution I've opened Homebrew/homebrew-core#69849 about this issue.

**Comment by veeragoni** [comment]
Like you said its looking at the wrong directory. Did you solve it by putting right key value in Caddy path? For example exposing a service via 8082,
instead of using caddy directly,

    <array>
      <string>caddy</string>
      <string>reverse-proxy</string>
      <string>--from</string>
      <string>localhost:8082</string>
      <string>--to</string>
      <string>localhost:8081</string>
    </array>


I would use this to specify the homebrew installation path by with `which caddy` where `/opt/homebrew/bin/` is the installation path for caddy 
```
<array>
      <string>/opt/homebrew/bin/caddy</string>
      <string>reverse-proxy</string>
      <string>--from</string>
      <string>localhost:8082</string>
      <string>--to</string>
      <string>localhost:8081</string>
 </array>

**Reply by veeragoni** [reply]
make sure to `launchctl unload <path of plist file>` and then `launchctl load <path of plist file>` once you set the right path
